drivers/bluetooth/hci/hci_sf32lb.c

```diff
@@ -0,0 +1,613 @@
+	return 0;
+}
+
+static void mbox_sf32lb_isr(const struct device *dev)
+{
+	LOG_DBG("%s: %p", __func__, (void *)dev);
+	LCPU2HCPU_IRQHandler();
+}
+
+static void zbt_config_mailbox(const struct device *dev)
+{
+	ipc_queue_cfg_t q_cfg;
+	struct bt_sf32lb_data *hci = dev->data;
+
+#ifndef CONFIG_SIFLI_LXT_DISABLE
+	/* Enable LXT32K */
+	HAL_PMU_EnableXTAL32();
+	if (HAL_PMU_LXTReady() != HAL_OK) {
+		HAL_ASSERT(0);
+	}
+	/* RTC/GTIME/LPTIME Using same low power clock source */
+	HAL_RTC_ENABLE_LXT();
+#endif
+	HAL_HPAON_StartGTimer();
+
+	k_sem_init(&(hci->sem), 0, 1);
+
+	q_cfg.qid = 0;
+	q_cfg.tx_buf_size = HCPU2LCPU_MB_CH1_BUF_SIZE;
+	q_cfg.tx_buf_addr = HCPU2LCPU_MB_CH1_BUF_START_ADDR;
+	q_cfg.tx_buf_addr_alias = HCPU_ADDR_2_LCPU_ADDR(HCPU2LCPU_MB_CH1_BUF_START_ADDR);
+
+#ifndef SF32LB52X
+	/* Config IPC queue. */
+	q_cfg.rx_buf_addr = LCPU_ADDR_2_HCPU_ADDR(LCPU2HCPU_MB_CH1_BUF_START_ADDR);
+#else
+	uint8_t rev_id = __HAL_SYSCFG_GET_REVID();
+
+	if (rev_id < HAL_CHIP_REV_ID_A4) {
+		q_cfg.rx_buf_addr = LCPU_ADDR_2_HCPU_ADDR(LCPU2HCPU_MB_CH1_BUF_START_ADDR);
+	} else {
+		q_cfg.rx_buf_addr = LCPU_ADDR_2_HCPU_ADDR(LCPU2HCPU_MB_CH1_BUF_REV_B_START_ADDR);
+	}
+#endif
+
+	q_cfg.rx_ind = NULL;
+	q_cfg.user_data = 0;
+
+	if (q_cfg.rx_ind == NULL) {
+		q_cfg.rx_ind = mbox_rx_ind;
+	}
+
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), mbox_sf32lb_isr,
+		    DEVICE_DT_INST_GET(0), 0);
+	hci->ipc_port = ipc_queue_init(&q_cfg);
+	__ASSERT(IPC_QUEUE_INVALID_HANDLE != hci->ipc_port, "Invalid Handle");
+	if (ipc_queue_open(hci->ipc_port)) {
+		__ASSERT(0, "Could not open IPC");
+	}
+}
+
+void sf32lb_hci_trace(uint8_t type, const uint8_t *data, uint16_t len, uint8_t direction)
+{
+	/*TODO: Add HCI trace functions*/
+}
+
+static inline void hci_get_type(const struct device *dev)
+{
+	struct bt_sf32lb_data *hci = dev->data;
+
+	/* Get packet type */
+	if (ipc_queue_read(hci->ipc_port, &hci->rx.type, 1) != 1) {
+		LOG_WRN("Unable to read H:4 packet type");
+		hci->rx.type = BT_HCI_H4_NONE;
+		return;
+	}
+
+	switch (hci->rx.type) {
+	case BT_HCI_H4_EVT:
+		hci->rx.remaining = sizeof(hci->rx.evt);
+		hci->rx.hdr_len = hci->rx.remaining;
+		break;
+	case BT_HCI_H4_ACL:
+		hci->rx.remaining = sizeof(hci->rx.acl);
+		hci->rx.hdr_len = hci->rx.remaining;
+		break;
+#if defined(CONFIG_BT_CLASSIC)
+	case BT_HCI_H4_SCO:
+		hci->rx.remaining = sizeof(hci->rx.sco);
+		hci->rx.hdr_len = hci->rx.remaining;
+		break;
+#endif
+#if defined(CONFIG_BT_ISO)
+	case BT_HCI_H4_ISO:
+		hci->rx.remaining = sizeof(hci->rx.iso);
+		hci->rx.hdr_len = hci->rx.remaining;
+		break;
+#endif
+	default:
+		goto error;
+	}
+	return;
+error:
+	LOG_ERR("Unknown HCI type 0x%02x", hci->rx.type);
+	hci->rx.type = BT_HCI_H4_NONE;
+}
+
+static void hci_read_hdr(const struct device *dev)
+{
+	struct bt_sf32lb_data *hci = dev->data;
+	int bytes_read = hci->rx.hdr_len - hci->rx.remaining;
+	int ret;
+
+	ret = ipc_queue_read(hci->ipc_port, hci->rx.hdr + bytes_read, hci->rx.remaining);
```

Comment:
Where is the implementation for this `ipc_queue_read()` function? It doesn't look like it's coming from Zephyr's IPC subsystem (which owns the `ipc_*` namespace in Zephyr), so this may be something that needs fixing.

Another question: what are the semantics of this API? Are they really identical to UART in that the transport is in practice a stream-based one which can give you partial reads instead of e.g. one HCI frame at a time? If there are differences compared to UART, then your reuse of all of the `h4.c` logic may not be appropriate.